pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -871,8 +849,12 @@ public int numMessagesInQueue() {
     public void redeliverUnacknowledgedMessages() {
         ClientCnx cnx = cnx();
         if (isConnected() && cnx.getRemoteEndpointProtocolVersion() >= ProtocolVersion.v2.getNumber()) {
-            if (unAckedMessageTracker != null) {
+            synchronized (this) {
+                int currentSize = incomingMessages.size();
+                incomingMessages.clear();
+                availablePermits.addAndGet(currentSize);
```

Comment:
Usecase:
1. consumer has prefetch queue size = 10 and at connection-opened consumer received all 10 messages
2. consumer is slow to ack-messages : and all messages have been cleared from ```incomingMessages``` and broker will receive ```RedeliverUnacknowledgedMessages``` command.
3. according to broker: consumer doesn't have permit
4. then will client be blocked on call: ```consumer.internalReceive()``` as there is no message on ```incomingMessages```